winrm/winrm.go

```diff
@@ -0,0 +1,271 @@
+	// - pass only client cert and key
+	// - 30 sec timeout
+	if ca := CACert(); ca != nil {
+		endpoint = winrm.NewEndpoint(host, httpsPort, true, false, ca, ClientCert(), ClientKey(), 30*time.Second)
+	} else {
+		endpoint = winrm.NewEndpoint(host, httpsPort, true, true, nil, ClientCert(), ClientKey(), 30*time.Second)
+		logger.Warningf("Skipping CA server verification, using the --insecure option")
+	}
+	logger.Debugf("Creating WinRM secure clinet connection on %s %d", host, httpsPort)
+	// if we don't have defined user
+	// use the default administrator on windows machine
+	if user == "" {
+		user = defaultWinndowsUser
+	}
+
+	cli.conn, cli.err = winrm.NewClient(endpoint, user, "")
+	if cli.err != nil {
+		cli.err = errors.Annotatef(cli.err, "Connot create WinRM secure client conn")
+	}
+	return cli
+}
+
+// GetPasswd callback for different semantics that a client
+// could use for secure authentication
+type GetPasswd func() (string, error)
+
+// TTYGetPasswd will be valid if it's used from a valid TTY input,
+// This can be used in conjunction with UseClient
+func TTYGetPasswd() (string, error) {
+	// very important that os.Stdin.Fd() needs to be a valid tty.
+	// if the file descriptor dosen't point to a terminal this will fail
+	// with the message inappropriate ioctl for device
+	pass, err := terminal.ReadPassword(int(os.Stdin.Fd()))
+	if err != nil {
+		return "", errors.Annotatef(err, "Can't retrive password from terminal")
+	}
+	return string(pass), nil
+}
+
+// NewClient binds new winrm client for initiating connections with the winrm listener
+// Note: use this function before winrm.LoadClientKeys()
+func NewClient(user, host string, gp GetPasswd, stdout io.Writer) *Client {
+	cli := &Client{}
+	winrm.DefaultParameters.TransportDecorator = nil
+
+	var endpoint *winrm.Endpoint
+	if ca := CACert(); ca != nil {
+		endpoint = winrm.NewEndpoint(host, httpPort, false, false, ca, nil, nil, 30*time.Second)
+	} else {
+		endpoint = winrm.NewEndpoint(host, httpPort, false, true, nil, nil, nil, 30*time.Second)
+		logger.Warningf("Skipping CA server verification, using the --insecure option")
+	}
+
+	logger.Debugf("Creating WinRM clinet connection on %s port %d", host, httpPort)
+
+	// if we don't have defined user
+	// use the default administrator on windows machine
+	if user == "" {
+		user = defaultWinndowsUser
+	}
+
+	if gp != nil { // get password from a given way
+		fmt.Fprintf(stdout, "[Winrm] >> ")
+		cli.pass, cli.err = gp()
+		if cli.err != nil {
+			return cli
+		}
+		fmt.Fprintf(stdout, "\n")
+	}
+
+	cli.conn, cli.err = winrm.NewClient(endpoint, user, cli.pass)
+	if cli.err != nil {
+		cli.err = errors.Annotatef(cli.err, "Connot create WinRM client conn")
+	}
+
+	return cli
+}
+
+// Run command and direct output to stdout and errors to stderr
+// If the Run successfully executes the cmd it returns nil on call of Error()
+func (c *Client) Run(command string, stdout io.Writer, stderr io.Writer) {
```

Comment:
@axw if you look on the other PR(https://github.com/juju/juju/pull/6523) you can see I extensively use winrm.Run() after I initialise it in the ```InitAdministratorUser``` function, It would be a little unmaintainable to make NewClient() and Close() every time I run winrm.Run() into a function, I should not create more connections.We should just one connection for trying https auth vs password auth interactions for the entire provisioning process or other processes that will be used in the juju-core.